tests/app/forms/validation/test_optional_form_validator.py

```diff
@@ -1,94 +1,67 @@
-import unittest
-from unittest.mock import Mock
-
+import pytest
 from wtforms.validators import StopValidation
 
 from app.forms.validators import OptionalForm
 
 
-class TestOptionalFormValidator(unittest.TestCase):
-    def test_date_optional_empty(self):
-        validator = OptionalForm()
-
-        mock_form = Mock()
-
-        mock_day = Mock()
-        mock_month = Mock()
-        mock_year = Mock()
-
-        mock_day.raw_data = [""]
-        mock_month.raw_data = []
-        mock_year.raw_data = [""]
-
-        mock_form.__iter__ = Mock(return_value=iter([mock_day, mock_month, mock_year]))
-
-        mock_field = Mock()
-
-        with self.assertRaises(StopValidation) as ite:
-            validator(mock_form, mock_field)
-
-        self.assertEqual("", str(ite.exception))
-
-    def test_month_year_optional_month_empty(self):
-        validator = OptionalForm()
-
-        mock_form = Mock()
-
-        mock_month = Mock()
-        mock_year = Mock()
-
-        mock_month.raw_data = []
-        mock_year.raw_data = [""]
-
-        mock_form.__iter__ = Mock(return_value=iter([mock_month, mock_year]))
-
-        mock_field = Mock()
-
-        with self.assertRaises(StopValidation) as ite:
-            validator(mock_form, mock_field)
-
-        self.assertEqual("", str(ite.exception))
-
-    def test_date_optional_missing_day(self):
+@pytest.mark.parametrize(
+    "day_month_year",
+    (
+        [[""], [], [""]],
+        [[], [""]],
+    ),
+)
+def test_date_validator_day_month_year_invalid_raises_StopValidation(
+    day_month_year, mocker, mock_form, mock_field
+):
+    validator = OptionalForm()
 
-        validator = OptionalForm()
+    mock_day = mocker.MagicMock()
+    mock_month = mocker.MagicMock()
+    mock_year = mocker.MagicMock()
 
-        mock_form = Mock()
+    mock_month.raw_data = day_month_year[-2]
+    mock_year.raw_data = day_month_year[-1]
 
-        mock_day = Mock()
-        mock_month = Mock()
-        mock_year = Mock()
+    if len(day_month_year) == 3:
+        mock_day.raw_data = day_month_year[-3]
+        comp = [mock_day, mock_month, mock_year]
 
-        mock_day.raw_data = [""]
-        mock_month.raw_data = ["01"]
-        mock_year.raw_data = ["2015"]
+    elif len(day_month_year) == 2:
```

Comment:
could be just `else`  there are only 2 scenarios

Reply:
I like to do this if there's a specific condition being tested but since I've made it so then yeah